### ISSUE TITLE
Possible memory leak by Backup sidecar during delta snapshot 

### ISSUE THREAD

**Issue by nikhilsap** [gardener/etcd-backup-restore#102]
We did few performance based test in the cluster, we noticed memory consumption by backup side car during Delta snapshot is more. 
Even after delta snapshot, memory usage is high and it is not coming down.

![image](https://user-images.githubusercontent.com/41104107/51895895-425c6800-23d1-11e9-93ee-03759502e07e.png)

Here in this case, 
We have loaded the db using benchmark tool. The DB size was 512 mb and delta snapshot period is 5 mins. we can see that, after Delta snapshot (13:14) the memory used by backup raised to more then 3GB where as ETCD memory consumption is around 350mb.

Here is other metrics during the same load.

![image](https://user-images.githubusercontent.com/41104107/51897768-7df93100-23d5-11e9-993a-116e2bcce40b.png)

We have trigged the load around 13.09 and it has taken full snapshot at 13.20. During full snapshot , it has a small rise in memory for both etcd and backup sidecar. 

etcd-backup-restore version : 0.5.0-dev-pref-test-3-memory and perf-test-2-with-metrics

**Comment by swapnilgm** [comment]
@nikhilsap  Willl you please mention the etcd-backup-restore version or commit id  on which you observed the issue?

**Comment by nikhilsap** [comment]
@swapnilgm i have updated the issue with etcd-backup-restore version.

**Comment by databus23** [comment]
We are seeing quite a lot of backup container (version 0.4.1) constantly hitting their memory limit in CCEE. The usage graph seems to suggest its some sort of memory leak:
![image](https://user-images.githubusercontent.com/439223/52711817-11fcf800-2f93-11e9-91bf-98da3341f068.png)

**Comment by amshuman-kr** [comment]
@databus23's [PR#116](https://github.com/gardener/etcd-backup-restore/pull/116) seems to have nailed it. 🥇👏 

A comparison is below.
1. [https://github.com/gardener/etcd-backup-restore/tree/master/head](https://github.com/gardener/etcd-backup-restore/commit/b173f994ce8b1a3c9e3f97c049d2cf3b5c887658)
<img width="1241" alt="gcvis-only-delta-master" src="https://user-images.githubusercontent.com/32157254/52798471-00ecdd80-309e-11e9-9548-6c9319668ca8.png">

1. @swapnilgm`s [PR#99](https://github.com/gardener/etcd-backup-restore/pull/99) as of [https://github.com/swapnilgm/etcd-backup-restore/tree/memory-prof-2](https://github.com/swapnilgm/etcd-backup-restore/commit/4df5b6052bc4ef1f0d7a67ef452ad06729c8ccd5).
<img width="1219" alt="gcvis-only-delta-mem-prof-2" src="https://user-images.githubusercontent.com/32157254/52798591-414c5b80-309e-11e9-8f62-217219980318.png">

1. @databus23's [PR#116](https://github.com/gardener/etcd-backup-restore/pull/116) as of [https://github.com/databus23/etcd-backup-restore/tree/fix-memory-leak](https://github.com/databus23/etcd-backup-restore/commit/307927d50748c14fd272f752a422c56fb02c2826).
<img width="1228" alt="gcvis-only-delta-fix-memory-leak" src="https://user-images.githubusercontent.com/32157254/52798822-b324a500-309e-11e9-8cc6-c4bc62c9aa24.png">

**Comment by amshuman-kr** [comment]
As a follow up, there is still a problem with full snapshot. It is not a leak anymore but full snapshot seems to require memory proportionate to the database size. See below. The spikes (at least the first two) seem to be from full snapshots. The measurement was done on @databus23's [PR#116](https://github.com/gardener/etcd-backup-restore/pull/116) as of [https://github.com/databus23/etcd-backup-restore/tree/fix-memory-leak](https://github.com/databus23/etcd-backup-restore/commit/307927d50748c14fd272f752a422c56fb02c2826).

<img width="1225" alt="gcvis-full-and-delta-fix-memory-leak" src="https://user-images.githubusercontent.com/32157254/52835020-ad64a900-310a-11e9-9d1d-36a76a33a8e4.png">

**Comment by nikhilsap** [comment]
We used PR https://github.com/gardener/etcd-backup-restore/pull/116 to check the memory consumption by etcd during snapshots.

we tried with 2 different DB sizes. (512mb and 2gb)

**Case 1 : 512 mb** 

![image](https://user-images.githubusercontent.com/41104107/52844902-b9149780-312b-11e9-81e7-7a9945619e92.png)


Here we can see, during delta snapshot memory consumption is there , but soon after the Full snapshot, it came down. Garbage Collection happened and Its not a leak anymore.

**Case 2 : 2 Gb**

![image](https://user-images.githubusercontent.com/41104107/52845170-6c7d8c00-312c-11e9-8d4d-c5ce6aaf6a17.png)

We can see the same thing here. During delta snapshot it consumes high memory. 
But here in 2 GB size, it took more time for memory to come down approx.9 mins. 
 Its not a leak anymore.
